### PR TITLE
core-services/ci-secret-generator: Document GCP XPN service account rotation

### PR DESCRIPTION
**Pull request by wking**
Like de19d42651 (#11910), but for XPN (shared VPC) jobs.  I'm also fixing a `delelete` typo from 09689a82a1 (#9094).

If you happen to blow away the GCP XPN service account used for provisioning CI jobs, you can recover following this procedure.  If you don't have gcloud installed locally, you can click "Activate Cloud Shell" [here][1].  Once you have gcloud in the openshift-gce-devel-ci project, I extracted the existing roles with:

```console
$ gcloud --format=json projects get-iam-policy openshift-gce-devel-ci | jq -r '.bindings[] | .role as $role | .members[] | select(. | contains("do-not-delelete-ci-xpn@")) | $role'
roles/compute.instanceAdmin.v1
roles/compute.networkAdmin
roles/compute.securityAdmin
roles/iam.serviceAccountUser
roles/storage.admin
```

To create a new account with those roles, run:

```sh
gcloud iam service-accounts create do-not-delete-ci-xpn --description='Credentials for creating GCP XPN (shared VPC) clusters in CI jobs.' --display-name='XPN (shared VPC) CI provisioner'
for ROLE in compute.instanceAdmin.v1 compute.networkAdmin compute.securityAdmin iam.serviceAccountUser storage.admin
do
  gcloud projects add-iam-policy-binding openshift-gce-devel-ci --member=serviceAccount:do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com --role="roles/${ROLE}"
done
```

Then the `keys create` I'm adding here will create a new key and push it into BitWarden.

You may also wish to revoke previous keys with:

```console
$ gcloud iam service-accounts keys list --iam-account do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com
KEY_ID         CREATED_AT            EXPIRES_AT
...redacted... 2020-09-16T20:23:20Z  9999-12-31T23:59:59Z
...more                                entries...
$ gcloud iam service-accounts keys delete $KEY_ID_TO_DELETE --iam-account do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com
```

or remove the previous service account with:

```console
$ gcloud iam service-accounts delete do-not-delelete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com
```

/assign @jstuever

CC @stevekuznetsov

[1]: https://console.cloud.google.com/iam-admin/iam?organizationId=54643501348&project=openshift-gce-devel-ci

### CODE REVIEW

**Comment by jstuever** [comment]
I have added the new SA to the host project as documented

```
#   - ${COMPUTE_SERVICE_ACCOUNT} (defined below)
#     - Compute Network User on both XPN Subnets (https://console.cloud.google.com/networking/xpn)
#   - ${CONTROL_SERVICE_ACCOUNT} (defined below)
#     - Compute Network User on both XPN Subnets (https://console.cloud.google.com/networking/xpn)
#     - Compute Network Viewer in HOST PROJECT
```

**Comment by wking** [comment]
> I have added the new SA to the host project as documented...

This is beyond me.  Is it something I should document in the commit message for future folks attempting to rotate this credential?

**Comment by wking** [comment]
@alvaroaleman, don't leave yet ;).  We need someone from DPTP to approve the changes to that file, and then run it to create a key and ingest that key into BitWarden.

**Comment by jstuever** [comment]
gcloud --project=openshift-dev-installer projects add-iam-policy-binding openshift-dev-installer --member "serviceAccount:do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com" --role "roles/compute.networkViewer"
gcloud --project=openshift-dev-installer compute networks subnets add-iam-policy-binding "installer-shared-vpc-subnet-1" --member "serviceAccount:do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com" --role "roles/compute.networkUser"
gcloud --project=openshift-dev-installer compute networks subnets add-iam-policy-binding "installer-shared-vpc-subnet-2" --member "serviceAccount:do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com" --role "roles/compute.networkUser"

File: core-services/ci-secret-generator/_config.yaml [hunk 1]
```diff
@@ -70,6 +70,10 @@
   attachments:
   - cmd: gcloud iam service-accounts keys create /dev/stdout --iam-account do-not-delete-ci-provisioner@openshift-gce-devel-ci.iam.gserviceaccount.com
     name: credentials.json
+- item_name: jenkins-ci-provisioner
+  attachments:
+  - cmd: gcloud iam service-accounts keys create /dev/stdout --iam-account do-not-delete-ci-xpn@openshift-gce-devel-ci.iam.gserviceaccount.com
+    name: credentials.json
```
Comment (jstuever) [inline core-services/ci-secret-generator/_config.yaml:76]: While this is "better", it isn't absolutely required; We can use `do-not-delete-ci-provisioner` for the installer to deploy the cluster while using `do-not-delete-ci-xpn` to attach to the gcp instances for "day 2" operations.

**Comment by jstuever** [comment]
/uncc @abhinavdahiya 
I don't believe he is an approver of `core-services/ci-secret-generator/_config.yaml`... need someone else here.

File: core-services/ci-secret-generator/_config.yaml (hunk 1 above)
Comment (wking) [reply core-services/ci-secret-generator/_config.yaml:76]: I am a bit fuzzy on how the XPN secret gets passed in.  Looks like the XPN jobs get the vanilla `gcp` cluster profile, and the XPN secrets are not mentioned in `ci-operator/SECRETS.md`.  Can you explain where the key is currently stored in BitWarden and how it flows to the workflow steps that need it?

File: core-services/ci-secret-generator/_config.yaml [hunk 2]
```diff
@@ -70,6 +70,10 @@
   attachments:
   - cmd: gcloud iam service-accounts keys create /dev/stdout --iam-account do-not-delete-ci-provisioner@openshift-gce-devel-ci.iam.gserviceaccount.com
     name: credentials.json
+- item_name: jenkins-ci-provisioner
```
Comment (wking) [inline core-services/ci-secret-generator/_config.yaml:73]: This value is almost certainly wrong.  We don't want to clobber the usual CI provisioner secret.

File: core-services/ci-secret-generator/_config.yaml (hunk 1 above)
Comment (wking) [reply core-services/ci-secret-generator/_config.yaml:76]: @jstuever : bump :).  Can you clarify the current secret flow^?

/assign @jstuever

Comment (jstuever) [reply core-services/ci-secret-generator/_config.yaml:76]: It doesn't. The XPN installer currently uses do-not-delete-ci-provisioner for provisioning the initial just like the rest of CI. The do-not-delete-ci-xpn attached to bootstrap, controlPlane, and compute instances where they can then use it to do post-provisioning tasks such as creating ingress lbs.

Comment (wking) [reply core-services/ci-secret-generator/_config.yaml:76]: How does the cluster access the creds?  Does the ci-operator pod mount a secret with a token, or...?

Comment (jstuever) [reply core-services/ci-secret-generator/_config.yaml:76]: XPN is only available in UPI. The service accounts are attached to the instances in the UPI template[1], which gives the instance the access granted to the service accounts.

[1] https://github.com/openshift/installer/blob/9be316bc89f560653e3ad07cde7a716bf938cf6d/upi/gcp/05_control_plane.py#L26

**Comment by wking** [comment]
Rebased around #14437 with ad3130c2be -> 0951a955ca.  I still need to know what the BitWarden location is for this secret.  I'm also confused by #14437's idempotent comment, because how can secret rotation possibly be idempotent?

**Pull request closed without merging**